### DISCUSSION TITLE
New plugin hooks

### DISCUSSION THREAD

**Discussion by Rem0o** [Rem0o/FanControl.Releases#1130] [Ideas]
Hi everybody, Rem0o here.

As of right now, the plugin interface only allows one to inject sensors. However, I plan to add a new interface to do much more.

For example, lots of people asked for different logics that allows to externally change config, either with hot keys, time of day and so on. That made me think I could expose a function to show available configs and an other to load a config. 

I got a few other ideas:

* Force show main window
* Read all sensors in FanControl
* Override a control

Combining all the ideas from above, one could make a plugin that monitors a temperature, and show the main window with a message if it gets too hot. It could also force all fans to go 100%, and so on. There are a lot of possibilities.

What do you guys think?

Discuss.

( people mentioned here either made plugins or gave me the ideas above )
@fu-raz @Mourdraug  @iJacks1980 @jmarucha @medevil84 @tuber84 @thewarrrr @Spuner

**Comment by medevil84** [comment]
Great idea! I would also add:
- Custom config page in the main settings (maybe as a tab?).
- Ability to show notifications

For example, with those additions, the aquacomputer highflownext plugin could force all the fan and pumps to 100% if the water temps goes above 57°C (water should not exceed 60°C temp), notify the user of what's happening via a toast notification and allow this kind of "alarm" to be deactivated/activated in the config page...

**Comment by ghost** [comment]
Love the idea, especially for certain devices such as the Corsair PRO/CORE doing one request for data rather than one for each sensor.

Would it also be possible to have more settings per plugin?  For example I've got a user that is struggling to get accurate temperature reads from his Commander PRO - turns out he wants greater precision than a 1 second data read.  I could setup a background worker to do this (as these devices just report when read), but right now I don't think it's possible to display the extra data - unless I am missing something really obvious ;)  I know this would be a lot more complex on the FanControl side.

Would it also be a possibility to have noisy notifications?  I like the idea of forcing the main window to show, but this could also cause problems with some (badly written) full screen programs.  So making the program scream (opt in of course) could be an option to let a user know that something is wrong, without potentially crashing programs - looking at you Destiny 2!

**Comment by jmarucha** [comment]
Hi Rem0o!

I was on pause with the free software contributions for a while, but had an idea. Some controllers (corsair h115i) have no pump settings in terms of duty, but in named presets like "balanced", "quiet" and so on. It would be nice if FanControl allowed some kind of interface for that.

I know building the fan graph is done in numbers, but maybe like connecting it to percentage input and setting in i-face that 0-33% is quiet, 34-66 is balanced, and 100 is performance.

Bests,
Jan
